bundle/edu.gemini.catalog/src/main/scala/edu/gemini/catalog/votable/VoTableClient.scala

```diff
@@ -20,17 +21,95 @@ trait VoTableBackend {
+   * Instead we need a function that can traverse the cache finding matches to reuse the result
+   */
+  object QueryCache {
+
+    // Container used internally on the cache, the type definition leaks into the client but that helps performance
+    // Vector was selected as it gives effective constant time for prepend, patch and dropRight
+    type CacheContainer[K, V] = Vector[CacheEntry[K, V]]
+    // Function provided by the caller to check if a value is on the cache
+    type FindFunction[K, V] = (CacheContainer[K, V], K) => Option[(Int, V)]
+
+    // based on scalaz's Memo.memo function
+    private def cache[K, V](f: (K => V) => K => V): Cache[K, V] = new Cache[K, V] {
+      def apply(z: K => V) = f(z)
+    }
+
+    // Build an LRU cache with a function that will traverse the cache finding suitable queries
+    private def lruCache[K, V](a: CacheContainer[K, V], findByKey: FindFunction[K, V], maxSize: Int = 100): Cache[K, V] = {
+      // Access needs to be synchronized
+      var m = a
+
+      cache[K, V](f => k => {
+        val value = a.synchronized {
+          val r = findByKey(m, k)
+          // Put it at the head if found
+          r.foreach {
+            case (pos, x) =>
+              // Remove it from its current position
+              m = m.patch(pos, Nil, 1)
+              // Move it to the front
+              m = CacheEntry(k, x) +: m
+          }
+          r.map(_._2)
+        }
+        value.getOrElse {
+          val r = f(k)
+          // Using smaller locks increases the chances to make multiple calls to f(k) but
+          // calling f(k) inside the lock notably reduces concurrency doing remote queries
+          a.synchronized {
+            // prepend the result at the beginning
+            m = CacheEntry(k, r) +: m
+            // Keep size constrained
+            if (m.size > maxSize) {
+              m = m.dropRight(1)
+            }
+          }
+          r
+        }
+      })
+    }
+
+    /**
+     * Builds a cache with a contains function to find cache hits
+     */
+    def buildCache[K, V](contains: FindFunction[K, V], maxSize: Int = 100) = lruCache(Vector.empty, contains, maxSize)
+    
+  }
+
+  private val cache:Cache[SearchKey, QueryResult] = {
+    // Find if a search is already in the cache
+    // Note that this assumes all catalogues give the same result for a given query
+    def contains(a: QueryCache.CacheContainer[SearchKey, QueryResult], k: SearchKey):Option[(Int, QueryResult)] = {
+      @tailrec
+      def go(a: QueryCache.CacheContainer[SearchKey, QueryResult], pos: Int):Option[(Int, QueryResult)] = {
+        a match {
+          case x if x.isEmpty                                                    => None
+          case x +: _ if x.k.query == k.query || x.k.query.isSuperSetOf(k.query) => Some((pos, x.v))
+          case x +: xs                                                           => go(a.tail, pos + 1)
+        }
```

Comment:
Destructuring a `Vector` in this way is actually quite expensive because it must construct a new instance for the tail. You might do better indexing into the vector directly:

```scala
def go(pos: Int): Option[(Int, QueryResult)] = {
  a.lift(pos) match {
    case Some(x) if x.k.query == k.query || x.k.query.isSuperSetOf(k.query) => Some((pos, x.v))
    case _ => go(pos + 1)
  }
```